app/src/main/java/net/gini/android/gvlexample/gini/SingleDocumentAnalyzer.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    public void cancelAnalysis() {
+        LOG.debug("Canceling analysis");
+        if (mAnalyzer != null) {
+            mAnalyzer.cancel();
+            LOG.debug("Analysis cancelled");
+        } else {
+            LOG.debug("No running analysis to cancel");
+        }
+    }
+
+    public void removeListener() {
+        LOG.debug("Removing listener");
+        if (mAnalyzer != null) {
+            mAnalyzer.setListener(null);
+            LOG.debug("Listener removed");
+        } else {
+            LOG.debug("No running analysis to remove listener from");
+        }
+    }
+
+    public void setListener(final DocumentAnalysisListener listener) {
+        LOG.debug("Setting listener");
+        if (mAnalyzer != null) {
+            mAnalyzer.setListener(new DocumentAnalysisListener() {
+                @Override
+                public void onExtractionsReceived(Map<String, SpecificExtraction> extractions) {
+                    logExtractions(extractions);
+                    listener.onExtractionsReceived(extractions);
+                }
+
+                @Override
+                public void onException(Exception exception) {
+                    LOG.error("Analysis failed", exception);
+                    listener.onException(exception);
+                }
+            });
+            LOG.debug("Listener set");
+        } else {
+            LOG.debug("No running analysis to set listener on");
+        }
+    }
+
+    public net.gini.android.models.Document getGiniApiDocument() {
+        LOG.debug("Getting Gini Api Document");
+        if (mAnalyzer != null){
+            if(mAnalyzer.isCompleted()) {
+                if (mAnalyzer.getGiniApiDocument() == null) {
+                    LOG.debug("No Gini Api Document Available: analysis failed");
+                }
+                return mAnalyzer.getGiniApiDocument();
+            } else {
+                LOG.debug("No Gini Api Document Available: analysis in progress");
+            }
+        } else {
+            LOG.debug("No Gini Api Document Available: no analysis started");
+        }
+        return null;
+    }
+
+    public interface DocumentAnalysisListener {
+        void onExtractionsReceived(Map<String, SpecificExtraction> extractions);
+
+        void onException(Exception exception);
+    }
+
+    private void logExtractions(Map<String, SpecificExtraction> extractions) {
+        LOG.debug("Extractions received:");
+        StringBuilder stringBuilder = new StringBuilder();
+        boolean first = true;
+        for (String extraction : extractions.keySet()) {
+            if (!first) {
+                stringBuilder.append("\n");
+            }
+            stringBuilder.append(extraction);
+            stringBuilder.append(" = ");
+            stringBuilder.append(extractions.get(extraction).getValue());
+            first = false;
+        }
+        LOG.debug(stringBuilder.toString());
+    }
+
+    private class Analyzer {
```

Comment:
This is probably a copy-paste issue, but I think this must be in a different class 😅

Reply:
I deliberately kept this here to have everything in one file in the GVL example apps. I'll clean it up a bit! 👍